sdk/aqueduct/utils/naming.py

```diff
@@ -0,0 +1,101 @@
+from typing import List, Optional, Tuple, Union
+
+from aqueduct.models.dag import DAG
+from aqueduct.utils.dag_deltas import RemoveOperatorDelta, apply_deltas_to_dag
+
+
+def _artifact_name_from_op_name(op_name: str) -> str:
+    return op_name + " artifact"
+
+
+def _construct_default_output_artifact_names_from_op(op_name: str, num_outputs: int) -> List[str]:
+    """The default artifact naming policy is "<op_name> artifact (<optional counter>)".
+
+    In the multi-output case, we deduplicate each, starting the counter at 1.
+    """
+    artifact_names = [_artifact_name_from_op_name(op_name)]
+    if num_outputs == 1:
+        return artifact_names
+
+    artifact_names += [
+        _artifact_name_from_op_name(op_name) + " (%d)" % i for i in range(1, num_outputs)
+    ]
+    return artifact_names
+
+
+def resolve_op_and_artifact_names(
+    dag: DAG,
+    candidate_op_name: str,
+    overwrite_existing_op_name: bool,
+    candidate_artifact_names: Optional[Union[str, List[str]]] = None,
+    num_outputs: int = 1,
+    only_resolve_op_name: bool = False,
+) -> Tuple[str, List[str]]:
+    """Enforces our unique naming policy on operators and also their output artifacts.
+
+    Operator collisions are resolved. Artifact collisions are not tolerated.
+
+    Args:
+        dag:
+            The current dag.
+        candidate_op_name:
+            The operator name that we want to check and potentially deduplicate.
+        overwrite_existing_op_name:
+            If set, any operator collisions will be ignored. Otherwise, we will bump
+            the operator name until an available one is found.
+        candidate_artifact_names:
+            The custom output artifact names the caller wants to assign to the operator
+            outputs. If not set, we will use our default artifact naming scheme. Regardless,
+            if there already exist artifact's with the same name, we will error.
+        num_outputs:
+            Must be consistent with `candidate_artifact_names`, if set. The number of
+            output artifacts to expected.
+        only_resolve_op_name:
```

Comment:
Just reading the comment here, since we do have uniqueness constraint on artifact names, if we skip the validation won't it cause bad behavior?

Reply:
Yes. We only ever skip the artifact name validation in the load operator case, since there is no artifact. I can add that to the docstring.

Reply:
That'd be great.